### PR TITLE
feat(sdk-core): add function to verify wallet signatures for TSS

### PR DESCRIPTION
**Pull request by LeoKotschenreuther**
When we create TSS wallets, BitGo returns two signatures so that
the owners of the user and backup keys are able to verify that
there was no man in the middle meddling with the involved key shares
and gpg keys.
These signatures are signed with BitGo's GPG public key and cover the
following items:
* commonKeychain
* key ID of the user GPG key
* key ID of the backup GPG key
* the derived public portion of the BitGo to User key share
* the derived public portion of the BitGo to Backup key share

Ticket: BG-59151

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My code compiles correctly for both Node and Browser environments
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [x] The ticket or github issue was included in the commit message as a reference
- [x] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: modules/sdk-core/src/bitgo/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/tss/eddsa/eddsa.ts]: I made them mandatory after our Slack conversation

File: modules/sdk-core/src/bitgo/tss/eddsa/eddsa.ts
Comment (islamaminBitGo) [inline modules/sdk-core/src/bitgo/tss/eddsa/eddsa.ts:399]: please use the function crypto_scalarmult_ed25519_base_noclamp instead of crypto_sign_ed25519_sk_to_pk.

File: modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts
Comment (Logicwax) [inline modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts:683]: why the use of padding?  we don't currently do that on the HSM side of things, so this may cause mis-alignment?

File: modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts
Comment (Logicwax) [inline modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts:707]: is 'nock' a typo for 'mock'?

File: modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts:683]: I was following the example from the OVC:
https://github.com/BitGo/offline-vault-console/blob/master/frontend/src/pkg/bitgo/mpc-utils.ts#L46
I will remove the padding from here to match what happens in the HSM.

File: modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts:709]: You gotta ask that the authors of the `nock` package :D
We are calling these helper functions `nockSomething` because they use the `nock` library to setup HTTP level mocks.

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (islamaminBitGo) [inline modules/sdk-core/src/bitgo/tss/common.ts:89]: To me, this seems a little risky, is there a more theoretically robust way to check exactly the certification coming from bitgo. I was thinking if we can loop on the pairs (keyid, valid) and find the pair where keyid matches bitgo's keyid, then check if valid is true for that pair.

Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:89]: `.verifyPrimaryUser` is a openpgpjs function, with index 0 being the self-cert of the CA key, and index 1 being the cert of the sub-key (our proof in this context).

unless openpgp.js is bumped and they change this logic, then this should always be the case.    notice that `openpgp.Key` is passed in, so it will check against that public key.   

Are you making the argument that index 0 should be checked for validity?

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (Logicwax) [inline modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:60]: why are you using userKeyId for the backupKeyId?

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (Logicwax) [inline modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:70]: same as above

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/tss/common.ts:89]: I wrote this code so that the signature has to be done with the bitgo pub to be considered valid. If I also consider index 0, I could pass in the user/ backup pub instead of the bitgo pub and would still consider the signature valid.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:49]: I must have missed this when I removed the other locations with the padding.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:60]: Because we haven't built anything yet that allows a third party to store the backup key shares. So far the user always  creates and stores the backup key shares and reuses the user gpg key as the backup gpg key.
Once we actually implement code to allow for a third party KRS we can change this code.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:70]: I explained above.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts]: Removed the padding, I also checked for more places that I might have missed but didn't find any more.

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:89]: index 0 should only show that the self-cert of the CA key is valid (gpg always includes a self-certification of the top CA key).

So by passing in bitgoPub, both indexes should return as valid.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts]: Excellent point. I still think that we currently have some valid use cases where the user and backup gpg key are the same but I updated this function so that they are separate.

File: modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts
Comment (Logicwax) [inline modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts:724]: your passing in userKeyId for backupkeyId, which should fail because on line 720 you are passing in `params.backupGpgKey.publicKey`

File: modules/bitgo/test/v2/unit/tss/helpers.ts
Comment (Logicwax) [inline modules/bitgo/test/v2/unit/tss/helpers.ts:30]: nit: if these are armored strings, then rename to something like "userWalletSigArmored`

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (Logicwax) [inline modules/sdk-core/src/bitgo/tss/common.ts:118]: what about the `userSharePub` and `backupSharePub`?

Remember, from the HSM, there are no uSig proofs.   Only wallet sigs.    which are just public keys of the user and backup party, with attached certifications from the HSM pubkey.   And each of those certs has notation data attached.

File: modules/bitgo/test/v2/unit/tss/helpers.ts
Comment (Logicwax) [inline modules/bitgo/test/v2/unit/tss/helpers.ts:36]: nit: rename to something like 'mergedWalletKeys` or `mergedWalletSigs`.   because this is not a one key, but rather is merging two pgp keys into one armored blob.

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (Logicwax) [inline modules/sdk-core/src/bitgo/tss/common.ts:87]: My apologies for providing incorrect info here.    `.verifyPrimaryUser` doesn't interact with subkeys.   It only checks certifications of the main key.

the code you have below:
```
const verificationResult = await walletSignature.verifyPrimaryUser([bitgoPub]);
const isValid = verificationResult.some((result) => result.valid);
```
is fine.   As this will check that the wallet pgp pubkey has at least ONE of it's attached certifications from the bitgo hsm pubkey (which is all we care about).   I mistakenly led you astray by thinking we were dealing with u-value subkeys, which we are not.  sorry about that.   This is only dealing with pgp CA keys certifying other pgp CA keys, and attaching notifcation data.

Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:87]: I would remove the comment here regarding subkeys, as that doesnt apply here

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (Logicwax) [inline modules/sdk-core/src/bitgo/tss/common.ts:92]: i really don't see the need for this.   especially after verifying the behavior of the `.verifyPrimaryUser` function

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/tss/common.ts:118]: The motivation was that we can reuse this function when we implement almost the same for ecdsa. As far as I understand it the code in this common function will remain the same but checking the user/backup share pubs will be different depending on the curve.

Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:118]: ah, fair point.   thanks for the explanation.

File: modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts:724]: I bet the tests were passing because we are still reusing the same gpg key for the user and backup when we create TSS wallets in the SDK. I updated the function to pass the backupKeyId though.

File: modules/bitgo/test/v2/unit/tss/helpers.ts
Comment (LeoKotschenreuther) [reply modules/bitgo/test/v2/unit/tss/helpers.ts:30]: Renamed

File: modules/bitgo/test/v2/unit/tss/helpers.ts
Comment (LeoKotschenreuther) [reply modules/bitgo/test/v2/unit/tss/helpers.ts:36]: Done

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/tss/common.ts]: Comment is removed.

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: I still do. If I remove this as well but leave the rest of the code the same then the following test will fail:
https://github.com/BitGo/BitGoJS/blob/79da30b69fb7be7a87d27f2ca463e751c20dda04/modules/bitgo/test/v2/unit/tss/common.ts#L38-L47
What this test is really doing is that a man in the middle is using their own gpg key (key1) to produce a proof over the BitGo pub (key2). This test should fail and it only fails if I keep this line.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (Logicwax) [inline modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:115]: pass in the backupGpgKey.publicKey

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (Logicwax) [inline modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:191]: same as above

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: that's not a MitM attack, that's not using the SDK properly.

the point of `commonVerifyWalletSignature()` is to take a pubkey, such as walletSignature, and verify if the second argument (bitgoPub in this case) certifies that same pubkey.     

in your test, you have key1 certifying key2 (you produce a key2 pub with a cert attached from key1).    then you pass in this certification (by passing in key2 pub with any certs attached), and bitgoPub/certifying key set to key2.   You're essentially asking the function to make sure that key2 has a certifcation from....key2 (and not key1, like you would expect).    Considering that bitgoPub is baked into the SDK, and should always be passed into the function, the only way to abuse this would be to modify this code (which wouldn't be mitm, which happens with the communication between parties).

I....guess I can agree it's good to keep guard rails here.   but if you're that concerned with someone using this code incorrectly, then i would argue that passing in bitgoPub at all should be removed and instead use the global defined bitgo public key.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:115]: There is no `backupGpgKey.publicKey` in this function available. Do you want me to change the entire wallet creation flow so that we use separate gpg keys for the user and the backup even though it's the same party?

Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:115]: In that case, are you ok if I do this in a separate PR? This change is going to bleed into the ecdsa code as well so I'd rather keep this as a separate work item.

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: I wasn't aware that bitgoPub is baked into the SDK. So far I only saw code that pulls it once from BitGo's `/api/v1/client/constants` API and then caches it.

Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: Where can I find the baked in bitgoPub?

Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: wow, I thought it was.   I guess that is only OVC and on platform.

Then yeah, we *should* bake this into SDK then.   that would help solve MitM problems.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (Logicwax) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:115]: yeah, i'm fine with that, as long as we track this in a ticket and not forget about it.

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: testnet: https://github.com/BitGo/bitgo-microservices/blob/56dd0f77c6a9878c8c55798ad1b903d6ecba4753/packages/wallet-platform/config/env/testnet.js#L145-L163

production:  https://github.com/BitGo/bitgo-microservices/blob/56dd0f77c6a9878c8c55798ad1b903d6ecba4753/packages/wallet-platform/config/env/production.js#L241-L254

I never liked the idea of pulling these from a remote API, since we should have a trust anchor in the SDK itself.

Comment (islamaminBitGo) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: I think we should look at this function in isolation and think of what it is promising, which is a robust check that walletSig has a non-self certification that is signed by bitgoPub and that the certification contains notation data that matches the wallet data that all parties have received so far in the creation of the wallet. Therefore, although baking bitgoPub would be the safest thing to do, the code as is will be passing it as a parameter, so I believe this check should stay there, otherwise if at some point bitgo's Pub key is baked as a global variable, I agree then that the check would be useless.

**Review by islamaminBitGo — DISMISSED** [review]
LGTM, but please wait for Matt's approval as well.

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (islamaminBitGo) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: But also, I just realized this check protects against the case if a MiM intercepts the communication between WP and User,  then replaces walletSig with their own walletSig that is just HSM's public key without even any other certs attached to it. In that case, wouldn't checking walletSig using bitgoPub through verifyPrimaryUser return valid without the above check? Because walletSig.verifyPrimaryUser([bitgoPub]) would return valid being true for the self certification in walletSig since the attacker just passed bitgoPub in the communication.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (sachushaji) [inline modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:44]: NIT: can add error descriptions for these asserts

**Review by sachushaji — APPROVED** [review]

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: @islamaminBitGo well I'm making the argument that we SHOULD bake it in.  It's important to have a trust anchor on the SDK itself, as MitM attacks could occur (this is why ovc even includes it)

> But also, I just realized this check protects against the case if a MiM intercepts the communication between WP and User, then replaces walletSig with their own walletSig that is just HSM's public key without even any other certs attached to it.

But the whole point of walletsigs, is that you get two GPG pubkeys and (after verifying the HSM has blessed them with a cert), you then cross-check that those are the gpg keys you've already learned are the two gpg keys to use (one being yourself, if you are the user, and the other being the backup).   Not only do you check that the HSM has certified those two keys, but that the certification itself includes notation data that specifies the two keyID's AND the commonPub (and also u-values).    

So if an attacker swapped out one of the walletsigs / user pubkey with the HSM one, then you should _still_ have a failure because the keyIDs are not lining up (you already know the gpg key of yourself, and you already know the gpg key of the backup provider) so you are able to check.   So if this single `if` statement is making that big of a difference, then we have evidence that our checking of these other things are not being performed correctly.

Comment (islamaminBitGo) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: Ah, exactly so one thing we're missing here in this check is an initial check of whether the two keys in walletSigs are indeed user's and backup's. 

Reiterating here to see if this is one way to move forward: 
Add one check in the function that dissects walletSigs into userGpgKeyWithCertAdded and backupGpgKeyWithCertAdded:
(Assume we're user and do the same thing for backup)
Check:
userGpgKeyWithCertAdded has the same fingerprint as our own key's fingerprint
backupGpgKeyWithCertAdded has the same fingerprint as that of the pub key we hold for backup

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (islamaminBitGo) [inline modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:58]: https://github.com/BitGo/BitGoJS/pull/2920/files#r1025889093

@Logicwax so we probably want to add a check here fir walletSignatures[0] and walletSignatures[1] given the ^ discussion.

File: modules/sdk-core/src/bitgo/tss/common.ts
Comment (Logicwax) [reply modules/sdk-core/src/bitgo/tss/common.ts:90]: yup, you got it.

the commonKeychain should also be checked too, but I guess that's done in the other function.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (Logicwax) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:58]: yep

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:115]: Ticket created and DMed to you.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:191]: Ticket created for further work.

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:58]: Just pushed some new code with the requested checks

File: modules/bitgo/example/js/create-tss-wallet.js
Comment (LeoKotschenreuther) [reply modules/bitgo/example/js/create-tss-wallet.js]: Change pushed and I also added a TS example

**Review by sachushaji — APPROVED** [review]
Thanks for addressing the comments

**Review by islamaminBitGo — COMMENTED** [review]
LGTM, great work! will let Matt give the approval for a merge.

**Comment by LeoKotschenreuther** [comment]
Had to fix another merge conflict

File: modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts
Comment (Logicwax) [inline modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:60]: im not opposed to this check (it's a good sanity check), but what should *really* be checked is if `userKeyId` is equal to the `userKeyId=` in the notation data of the signatures from the HSM.    

https://github.com/BitGo/hsm-api/blob/44c9e8295407d6df411e547c63d38a07a9edc6e2/src/api/newsolanakey.js#L299-L314

That's the notation data.    This means that when the HSM certifies the user/backup gpg key, its signing packets that (in addition to standard pgp packets of certifying another key) _also_ include that data.

so just do what you did in `commonVerifyWalletSignature()` where you check against the raw notations.    `userKeyId !== Buffer.from(rawNotations[1].value).toString(),`

Comment (LeoKotschenreuther) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:60]: This function is (transitively) calling `commonVerifyWalletSignature` and passing the user and backup key ids. This means that it's already checking what you are suggesting.
It first calls `verifyWalletSignature` for each signature and passes the user and backup key ids:
https://github.com/BitGo/BitGoJS/blob/0e6840e4b9a89aea30e784e0acede2377937fe6c/modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts#L68-L86
`verifyWalletSignature` then calls `commonVerifyWalletSignature` and passes the same params in:
https://github.com/BitGo/BitGoJS/blob/0e6840e4b9a89aea30e784e0acede2377937fe6c/modules/sdk-core/src/bitgo/tss/eddsa/eddsa.ts#L413
The checks you are asking for are done here:
https://github.com/BitGo/BitGoJS/blob/0e6840e4b9a89aea30e784e0acede2377937fe6c/modules/sdk-core/src/bitgo/tss/common.ts#L111-L115

Comment (Logicwax) [reply modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts:60]: Thanks the explanation.   resolving this.

**Review by islamaminBitGo — APPROVED** [review]

**Pull request merged**
